src/lazybroadcasting.jl

```diff
@@ -37,9 +37,91 @@ getindex(B::BroadcastArray{<:Any,1}, kr::AbstractVector{<:Integer}) =
 
 copy(bc::Broadcasted{<:LazyArrayStyle}) = BroadcastArray(bc)
 
-# issue 16: sum(b, dims=(1,2,3)) faster than sum(b)
-Base._sum(b::BroadcastArray{T,N}, ::Colon) where {T,N} = first(Base._sum(b, ntuple(identity, N)))
-Base._prod(b::BroadcastArray{T,N}, ::Colon) where {T,N} = first(Base._prod(b, ntuple(identity, N)))
+# Replacement for #18.
+# Could extend this to other similar reductions in Base... or apply at lower level? 
+# for (fname, op) in [(:sum, :add_sum), (:prod, :mul_prod),
+#                     (:maximum, :max), (:minimum, :min),
+#                     (:all, :&),       (:any, :|)]
+function Base._sum(f, A::BroadcastArray, ::Colon)
+    bc = A.broadcasted
+    T = Broadcast.combine_eltypes(f ∘ bc.f, bc.args) 
+    out = zero(T)
+    @simd for I in eachindex(bc)
+        @inbounds out += f(bc[I])
+    end
+    out
+end
+function Base._prod(f, A::BroadcastArray, ::Colon)
+    bc = A.broadcasted
+    T = Broadcast.combine_eltypes(f ∘ bc.f, bc.args) 
+    out = one(T)
+    @simd for I in eachindex(bc)
+        @inbounds out *= f(bc[I])
+    end
+    out
+end
+
+# Macros for lazy broadcasting, #21 WIP
+# based on @dawbarton  https://discourse.julialang.org/t/19641/20
+# and @tkf            https://github.com/JuliaLang/julia/issues/19198#issuecomment-457967851
+# and @chethega      https://github.com/JuliaLang/julia/pull/30939
+
+export @lazy, @lazydot, @□, @⊡
+
+lazy(::Any) = error("function `lazy` must be called with a dot")
+struct LazyCast{T}
+    value::T
+end
+Broadcast.broadcasted(::typeof(lazy), x) = LazyCast(x)
+Broadcast.materialize(x::LazyCast) = BroadcastArray(x.value)
+
+lazyhelp = """
+    @lazy A .+ B     == @□ A .+ B
+    @lazydot A + B   == @⊡ A + B
+
+Macros for creating lazy `BroadcastArray`s: `@lazy` expects a broadcasting expression, 
+while `@lazydot` applies `@.` first. Short forms are typed `@\\square` and `@\\boxdot` 
+(or perhaps `@z` & `@ż` except that `ż` seems hard to enter at the REPL).
+"""
+
+@doc lazyhelp
+macro lazy(ex)
+    checkex(ex)
+    :( lazy.($(esc(ex))) )
```

Comment:
Isn't `esc(:( $lazy.($ex) ))` better?  It'd work better with macros inside `@lazy`.

Reply:
I never thought to use `$` for that, am new to macros. Both give `@macroexpand @~ @. A + 2` --> `:((LazyArrays.lazy).((+).(A, 2)))`, how do I see the difference?

Reply:
Ah, sorry, never mind.  I think I was mixing it with something else.